frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx

```diff
@@ -0,0 +1,414 @@
+  };
+};
+
+const subformFactory = (receiverType: string) => {
+  switch (receiverType) {
+    case 'pagerduty_configs':
+      return PagerDutyForm;
+    case 'webhook_configs':
+      return WebhookForm;
+    default:
+      return null;
+  }
+};
+
+function formReducer(formValues, action) {
+  if (action.type === 'setFormValues') {
+    return {
+      ...formValues,
+      ...action.payload,
+    };
+  }
+  // eslint-disable-next-line no-console
+  console.error('Unrecognized Alertmanager form reducer action', action);
+  return formValues;
+}
+
+const getRouteLabelsForEditor = (
+  isDefaultReceiver: boolean,
+  receiverToEdit: AlertmanagerReceiver,
+  allRoutes: AlertmanagerRoute[],
+): RouteEditorLabel[] => {
+  const routeLabels = convertReceiverRoutesToEditorLabels(receiverToEdit, allRoutes);
+  return !isDefaultReceiver && _.isEmpty(routeLabels)
+    ? [{ name: '', value: '', isRegex: false }]
+    : routeLabels;
+};
+
+const ReceiverBaseForm: React.FC<ReceiverBaseFormProps> = ({
+  obj: secret, // Secret "alertmanager-main" which contains alertmanager.yaml config
+  titleVerb,
+  saveButtonText,
+  editReceiverNamed,
+}) => {
+  const [errorMsg, setErrorMsg] = React.useState<string>();
+  const [inProgress, setInProgress] = React.useState<boolean>(false);
+  let config: AlertmanagerConfig;
+  if (!errorMsg) {
+    config = getAlertmanagerConfig(secret, setErrorMsg);
+  }
+
+  const INITIAL_STATE = {
+    receiverName: '',
+    receiverType: '',
+    routeLabels: [],
+    ...PagerDutyForm.getInitialValues(null),
+    ...WebhookForm.getInitialValues(null),
+  };
+
+  let receiverToEdit: AlertmanagerReceiver;
+  if (editReceiverNamed) {
+    receiverToEdit = _.find(_.get(config, 'receivers'), { name: editReceiverNamed });
+    INITIAL_STATE.receiverName = receiverToEdit ? receiverToEdit.name : '';
+    INITIAL_STATE.receiverType = _.find(_.keys(receiverToEdit), (key) =>
+      _.endsWith(key, '_configs'),
+    );
+    if (!_.isEmpty(INITIAL_STATE.receiverType)) {
+      const receiverConfig = receiverToEdit?.[INITIAL_STATE.receiverType]?.[0];
+      _.assign(
+        INITIAL_STATE,
+        subformFactory(INITIAL_STATE.receiverType).getInitialValues(receiverConfig),
+      );
+    }
+  }
+
+  const { route } = config || {};
+  const { receiver: defaultReceiver } = route || {}; // top level route.receiver is the default receiver for all alarms
+  // if no default receiver defined or editing the default receiver
+  const isDefaultReceiver = defaultReceiver === undefined || defaultReceiver === editReceiverNamed;
+
+  INITIAL_STATE.routeLabels = getRouteLabelsForEditor(
+    isDefaultReceiver,
+    receiverToEdit,
+    route.routes,
+  );
+
+  const [formValues, dispatchFormChange] = React.useReducer(formReducer, INITIAL_STATE);
+  const SubForm = subformFactory(formValues.receiverType);
+
+  const isFormInvalid =
+    _.isEmpty(formValues.receiverName) ||
+    _.isEmpty(formValues.receiverType) ||
```

Comment:
If these are strings, you don't need to use `_.isEmpty`, just `!formValues.receiverName`

Reply:
yeah, I had this fixed in large PR we closed, lost them in recreating this PR.  fixed